Project21/Project21/intlist.c

```diff
@@ -0,0 +1,212 @@
+
+
+
+/// <summary>Return the number of elements in list</summary>
+/// <param name="head"></param>
+/// <returns>Error code (if any)</returns>
+int Count (LinkedList* list) {
+   //checks if the list is deleted
+   if (list->deleted == true) return E_LIST_DOES_NOT_EXIST;
+   //if the list is not deleted, it iterates to get count
+   int count = 0;
+   Node* current = list->head;
+   while (current != NULL) {
+      count++;/// Increment count by 1
+      current = current->next;
+   }
+   return count;///Return the count of nodes
+}
+
+
+/// <summary>Delete the entire list</summary>
+/// <param name="head"></param>
+/// <returns>Error code (if any)</returns>
+int Delete (LinkedList* list) {
+   //checks if the list is deleted
+   if (list->deleted == true) return E_LIST_DOES_NOT_EXIST;
+   //Checks if the list is empty and deletes list
+   Node* current = list->head;
+   Node* next = NULL;
+   if (current == NULL) {
+      list->deleted = true;
+      return 0;
+   }
+   //if the list is not empty, it iterates to the end and deletes all nodes
+   while (current != NULL) {
+      next = current->next;
+      free (current);
+      current = next;
+   }
+   list->head = NULL;
+   list->deleted = true;
+   return 0;
+}
+
+/// <summary>Removes the first occurrence of specified value in list</summary>
+/// <param name="head"></param>
+/// <param name="data"></param>
+/// <returns>Error code (if any)</returns>
+int Remove (LinkedList* list, int data) {
+   // checks if the list is deleted
+   if (list->deleted == true) return E_LIST_DOES_NOT_EXIST;
+   //Checks if the list is empty
+   Node* temp = list->head;
+   Node* pre = NULL;
+   if (temp == NULL) return E_EMPTY_LIST;
+   //Check if the data in head(1st node) is to be removed
+   if (temp != NULL && temp->data == data) {
+      list->head = temp->next;
+      free (temp);
+      return 0;
+   }
+   //if the list is not empty, it iterates to remove the data at spec.index
+   while (temp != NULL && temp->data != data) {//if the list is not empty, it iterates to remove the data at spec. index
+      pre = temp;
+      temp = temp->next;
+   }
+   //checks if index exceeds list count
+   if (temp != NULL) {
+      pre->next = temp->next;
+      free (temp);
+      return 0;
+   }
+   else {
+      return E_VALUE_NOT_FOUND;
+   }
+}
+/// <summary>Removes the value at a specified index</summary>
+/// <param name="head"></param>
+/// <param name="index"></param>
+/// <returns>Error code (if any)</returns>
+int RemoveAt (LinkedList* list, int index) {
+   //checks if the list is deleted
+   if (list->deleted == true) return E_LIST_DOES_NOT_EXIST;
+   //Checks if the list is empty
+   Node* temp = list->head;
+   Node* pre = NULL;
+   if (temp == NULL) return E_EMPTY_LIST;
+   //Check if the data in head(1st node) is to be removed
+   if (index == 0) {
+      list->head = temp->next;
+      free (temp);
+      return 0;
+   }
+   //if the list is not empty, it iterates to remove the data at spec. index

```

Comment:
You should also check to see if the index values passed is valid before traversing. There is no guarantee that temp and pre will be null and not point to garbage values once you go beyond the last value.